keps/sig-storage/2644-honor-pv-reclaim-policy/README.md

```diff
@@ -0,0 +1,321 @@
+
+### Test Plan
+
+#### E2E tests
+An e2e test to exercise deletion of PV prior to deletion of PVC for a `Bound` PV-PVC pair.
+
+### Graduation Criteria
+
+#### Alpha
+* Feedback
+* Fix of the issue behind a feature flag.
+* Unit tests and e2e tests outlined in design proposal implemented.
+* Tests should be done with both CSI Migration disabled and enabled.
+
+#### Alpha -> Beta
+* Allow the Alpha fix to soak for one release.
+
+#### Beta -> GA
+* Deployed in production and have gone through at least one K8s upgrade.
+
+### Upgrade / Downgrade Strategy
+* Upgrade from old Kubernetes(1.22) to new Kubernetes(1.23) with `HonorPVReclaimPolicy` flag enabled
+
+In the above scenario, the upgrade will cause change in default behavior as described in the current KEP. Additionally,
+if there are PVs that have a valid associated PVC and deletion timestamp set, then a finalizer is added to the PV.
+
+* Downgrade from new Kubernetes(1.23) to old Kubernetes(1.22).
+
+In this case, there may be PVs with the deletion finalizer that the older Kubernetes does not remove. Such PVs will be in the API server forever unless if the user manually removes them.
+
+### Version Skew Strategy
+The fix is part of `kube-controller-manager` of `external-provisioner`.
+
+1. `kube-controller-manager` is upgraded, but `external-provisioner` is not:
+
+In this case the changed behavior is observed only for in-tree plugins. The drivers would still have the issue. Users
+will observe the new annotation on the PV, but it does not have any effect since the `external-provisioner` is not
+updated
+
+2. `external-provisioner` is upgraded but `kube-controller-manager`
+
+In this case the annotations are not added, the upgraded `external-provisioner` will have no effect, the issue would
+still exist.
+
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: `HonorPVReclaimPolicy`
+  - Components depending on the feature gate: kube-controller-manager, external-provisioner
+
+###### Does enabling the feature change any default behavior?
+
+Enabling the feature will delete the volume from underlying storage when the PV is deleted followed deleting the PVC for
+a bound PV-PVC pair where the PV reclaim policy is delete.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+Yes. Disabling the feature flag will continue previous behavior.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+Will pick the new behavior as described before.
+
+###### Are there any tests for feature enablement/disablement?
+
+There will be unit tests for the feature `HonorPVReclaimPolicy` enablement/disablement.
+
+### Scalability
+
+###### Will enabling / using this feature result in any new API calls?
+No.
```

Comment:
I think something will need to add the new finalizer to all existing PVs when the feature is enabled. Depending on amount of PVs, I expect some throttling on the API server.
This will be one-time sync.

Reply:
Fixed.